src/core/stdcpp/allocator.d

```diff
@@ -0,0 +1,54 @@
+/**
+ * D header file for interaction with C++ std::allocator.
+ *
+ * Copyright: Copyright (c) 2018 D Language Foundation
+ * License: Distributed under the
+ *      $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0).
+ *    (See accompanying file LICENSE)
+ * Authors:   Guillaume Chatelet
+ *            Manu Evans
+ * Source:    $(DRUNTIMESRC core/stdcpp/allocator.d)
+ */
+
+module core.stdcpp.allocator;
+
+extern(C++, "std"):
+
+/**
+ * Allocators are classes that define memory models to be used by some parts of
+ * the C++ Standard Library, and most specifically, by STL containers.
+ */
+extern(C++, class) struct allocator(T)
+{
+    static assert(!is(T == const), "The C++ Standard forbids containers of const elements because allocator!(const T) is ill-formed.");
+    static assert(!is(T == immutable), "immutable is not representable in C++");
+
+    alias value_type = T;
+
+    version (CppRuntime_Microsoft)
+    {
+        T* allocate(size_t count) nothrow @nogc;
+        // HACK: workaround to make `deallocate` link as a `T * const`
+        extern (D) private static string constHack(string name)
+        {
+            version (Win64)
+                enum sub = "AAXPE";
+            else
+                enum sub = "AEXPA";
+            foreach (i; 0 .. name.length - sub.length)
+                if (name[i .. i + sub.length] == sub[])
+                    return name[0 .. i + 3] ~ 'Q' ~ name[i + 4 .. $];
+            assert(false, "substitution string not found!");
+        }
+        pragma(mangle, constHack(deallocate.mangleof))
+        void deallocate(T* ptr, size_t count) nothrow @nogc;
```

Comment:
This is interesting: deallocate sets the mangling by retrieving and modifying its own mangling. It seems to works, but looks brittle.

Reply:
Yeah, I did not expect this to work.

Reply:
Actually, in my tests (by default with VS2013) the const hack is wrong. Linking succeeds without it.

Reply:
Wonderful! Do we have a version or `__tratits(getTargetInfo, "blah")` to detect what version of VS we are linking against?

Reply:
I retried with VS2015 and VS2017, these also pass without complaining about `deallocate` (without the const hack), but a missing generic delete in the array-test:
```
"c:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023/bin/Hostx64/x64/cl.exe" -c /Foallocator_c
pp.obj test\stdcpp\src\allocator.cpp /EHsc
Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27023.1 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

allocator.cpp

"c:\s\d\rainers\windows\bin\dmd.exe" -of=allocator.exe -m64 -conf= -Isrc -defaultlib=lib\druntime64.lib -main -unittest test\stdcp
p\src\allocator_test.d allocator_cpp.obj

allocator.exe

del allocator.exe allocator.obj allocator_cpp.obj
"c:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023/bin/Hostx64/x64/cl.exe" -c /Foarray_cpp.o
bj test\stdcpp\src\array.cpp /EHsc
Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27023.1 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

array.cpp

"c:\s\d\rainers\windows\bin\dmd.exe" -of=array.exe -m64 -conf= -Isrc -defaultlib=lib\druntime64.lib -main -unittest test\stdcpp\sr
c\array_test.d array_cpp.obj
array_cpp.obj : error LNK2019: unresolved external symbol "void __cdecl operator delete(void *,unsigned __int64)" (??3@YAXPEAX_K@Z
) referenced in function "public: virtual void * __cdecl std::_Generic_error_category::`scalar deleting destructor'(unsigned int)"
 (??_G_Generic_error_category@std@@UEAAPEAXI@Z)
array.exe : fatal error LNK1120: 1 unresolved externals
Error: linker exited with status 1120

--- errorlevel 1
```

Reply:
Yes this line is sensitive to VS version. I wrote this against VS2017, which is 'current', but we may need to be able to specify the target VS version (similar problem to the recent `-cppstd` debate.